docs/ext/remix_code_links.py

```diff
@@ -0,0 +1,53 @@
+import base64
+import docutils
+
+
+def insert_node_before(child, new_sibling):
+    assert child in child.parent.children
+
+    for position, node in enumerate(child.parent.children):
+        if node == child:
+            child.parent.insert(position, new_sibling)
+            break
+
+
+def remix_code_url(source_code):
+    # NOTE: base64 encoded data may contain +, = and / characters. Remix seems to handle them just
+    # fine without any escaping.
+    base64_encoded_source = base64.b64encode(source_code.encode('utf-8')).decode('ascii')
+    return f"https://remix.ethereum.org/?code={base64_encoded_source}"
```

Comment:
https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Reply:
Does remix has a version parameter still, or will it just find it out based on the pragma?

Reply:
Does remix has a version parameter still, or will it just find it out based on the pragma?

Reply:
That's a good point. Still, I'm not sure if we can do anything reasonable about that. I could make it not render the link if it's too long but the limit is not consistent across browser so I think it's better to keep it for those browsers that do support longer URLs.

We could ask Remix to provide a `POST` endpoint that would allow posting longer code.

We could also use a URL shortener at build time but that sounds complicated and brittle.

Reply:
> We could ask Remix to provide a POST endpoint that would allow posting longer code.

Can do that easily from the user side.

> We could also use a URL shortener at build time but that sounds complicated and brittle.

Just ask remix to support lib compression and/or ascii85 in the payload :)

Reply:
At the minimum you can output a warning on `make docs` if any snippet has an URL longer than 2000 bytes. We could then see if anything is affected.

Reply:
The "Voting" example seems to be the longest Solidity snippet currently. It's a bit under 8 kB. [The link that gets generated for it](https://remix.ethereum.org/?code=Ly8gU1BEWC1MaWNlbnNlLUlkZW50aWZpZXI6IEdQTC0zLjAKcHJhZ21hIHNvbGlkaXR5ID49MC43LjAgPDAuOS4wOwovLy8gQHRpdGxlIFZvdGluZyB3aXRoIGRlbGVnYXRpb24uCmNvbnRyYWN0IEJhbGxvdCB7CiAgICAvLyBUaGlzIGRlY2xhcmVzIGEgbmV3IGNvbXBsZXggdHlwZSB3aGljaCB3aWxsCiAgICAvLyBiZSB1c2VkIGZvciB2YXJpYWJsZXMgbGF0ZXIuCiAgICAvLyBJdCB3aWxsIHJlcHJlc2VudCBhIHNpbmdsZSB2b3Rlci4KICAgIHN0cnVjdCBWb3RlciB7CiAgICAgICAgdWludCB3ZWlnaHQ7IC8vIHdlaWdodCBpcyBhY2N1bXVsYXRlZCBieSBkZWxlZ2F0aW9uCiAgICAgICAgYm9vbCB2b3RlZDsgIC8vIGlmIHRydWUsIHRoYXQgcGVyc29uIGFscmVhZHkgdm90ZWQKICAgICAgICBhZGRyZXNzIGRlbGVnYXRlOyAvLyBwZXJzb24gZGVsZWdhdGVkIHRvCiAgICAgICAgdWludCB2b3RlOyAgIC8vIGluZGV4IG9mIHRoZSB2b3RlZCBwcm9wb3NhbAogICAgfQoKICAgIC8vIFRoaXMgaXMgYSB0eXBlIGZvciBhIHNpbmdsZSBwcm9wb3NhbC4KICAgIHN0cnVjdCBQcm9wb3NhbCB7CiAgICAgICAgYnl0ZXMzMiBuYW1lOyAgIC8vIHNob3J0IG5hbWUgKHVwIHRvIDMyIGJ5dGVzKQogICAgICAgIHVpbnQgdm90ZUNvdW50OyAvLyBudW1iZXIgb2YgYWNjdW11bGF0ZWQgdm90ZXMKICAgIH0KCiAgICBhZGRyZXNzIHB1YmxpYyBjaGFpcnBlcnNvbjsKCiAgICAvLyBUaGlzIGRlY2xhcmVzIGEgc3RhdGUgdmFyaWFibGUgdGhhdAogICAgLy8gc3RvcmVzIGEgYFZvdGVyYCBzdHJ1Y3QgZm9yIGVhY2ggcG9zc2libGUgYWRkcmVzcy4KICAgIG1hcHBpbmcoYWRkcmVzcyA9PiBWb3RlcikgcHVibGljIHZvdGVyczsKCiAgICAvLyBBIGR5bmFtaWNhbGx5LXNpemVkIGFycmF5IG9mIGBQcm9wb3NhbGAgc3RydWN0cy4KICAgIFByb3Bvc2FsW10gcHVibGljIHByb3Bvc2FsczsKCiAgICAvLy8gQ3JlYXRlIGEgbmV3IGJhbGxvdCB0byBjaG9vc2Ugb25lIG9mIGBwcm9wb3NhbE5hbWVzYC4KICAgIGNvbnN0cnVjdG9yKGJ5dGVzMzJbXSBtZW1vcnkgcHJvcG9zYWxOYW1lcykgewogICAgICAgIGNoYWlycGVyc29uID0gbXNnLnNlbmRlcjsKICAgICAgICB2b3RlcnNbY2hhaXJwZXJzb25dLndlaWdodCA9IDE7CgogICAgICAgIC8vIEZvciBlYWNoIG9mIHRoZSBwcm92aWRlZCBwcm9wb3NhbCBuYW1lcywKICAgICAgICAvLyBjcmVhdGUgYSBuZXcgcHJvcG9zYWwgb2JqZWN0IGFuZCBhZGQgaXQKICAgICAgICAvLyB0byB0aGUgZW5kIG9mIHRoZSBhcnJheS4KICAgICAgICBmb3IgKHVpbnQgaSA9IDA7IGkgPCBwcm9wb3NhbE5hbWVzLmxlbmd0aDsgaSsrKSB7CiAgICAgICAgICAgIC8vIGBQcm9wb3NhbCh7Li4ufSlgIGNyZWF0ZXMgYSB0ZW1wb3JhcnkKICAgICAgICAgICAgLy8gUHJvcG9zYWwgb2JqZWN0IGFuZCBgcHJvcG9zYWxzLnB1c2goLi4uKWAKICAgICAgICAgICAgLy8gYXBwZW5kcyBpdCB0byB0aGUgZW5kIG9mIGBwcm9wb3NhbHNgLgogICAgICAgICAgICBwcm9wb3NhbHMucHVzaChQcm9wb3NhbCh7CiAgICAgICAgICAgICAgICBuYW1lOiBwcm9wb3NhbE5hbWVzW2ldLAogICAgICAgICAgICAgICAgdm90ZUNvdW50OiAwCiAgICAgICAgICAgIH0pKTsKICAgICAgICB9CiAgICB9CgogICAgLy8gR2l2ZSBgdm90ZXJgIHRoZSByaWdodCB0byB2b3RlIG9uIHRoaXMgYmFsbG90LgogICAgLy8gTWF5IG9ubHkgYmUgY2FsbGVkIGJ5IGBjaGFpcnBlcnNvbmAuCiAgICBmdW5jdGlvbiBnaXZlUmlnaHRUb1ZvdGUoYWRkcmVzcyB2b3RlcikgcHVibGljIHsKICAgICAgICAvLyBJZiB0aGUgZmlyc3QgYXJndW1lbnQgb2YgYHJlcXVpcmVgIGV2YWx1YXRlcwogICAgICAgIC8vIHRvIGBmYWxzZWAsIGV4ZWN1dGlvbiB0ZXJtaW5hdGVzIGFuZCBhbGwKICAgICAgICAvLyBjaGFuZ2VzIHRvIHRoZSBzdGF0ZSBhbmQgdG8gRXRoZXIgYmFsYW5jZXMKICAgICAgICAvLyBhcmUgcmV2ZXJ0ZWQuCiAgICAgICAgLy8gVGhpcyB1c2VkIHRvIGNvbnN1bWUgYWxsIGdhcyBpbiBvbGQgRVZNIHZlcnNpb25zLCBidXQKICAgICAgICAvLyBub3QgYW55bW9yZS4KICAgICAgICAvLyBJdCBpcyBvZnRlbiBhIGdvb2QgaWRlYSB0byB1c2UgYHJlcXVpcmVgIHRvIGNoZWNrIGlmCiAgICAgICAgLy8gZnVuY3Rpb25zIGFyZSBjYWxsZWQgY29ycmVjdGx5LgogICAgICAgIC8vIEFzIGEgc2Vjb25kIGFyZ3VtZW50LCB5b3UgY2FuIGFsc28gcHJvdmlkZSBhbgogICAgICAgIC8vIGV4cGxhbmF0aW9uIGFib3V0IHdoYXQgd2VudCB3cm9uZy4KICAgICAgICByZXF1aXJlKAogICAgICAgICAgICBtc2cuc2VuZGVyID09IGNoYWlycGVyc29uLAogICAgICAgICAgICAiT25seSBjaGFpcnBlcnNvbiBjYW4gZ2l2ZSByaWdodCB0byB2b3RlLiIKICAgICAgICApOwogICAgICAgIHJlcXVpcmUoCiAgICAgICAgICAgICF2b3RlcnNbdm90ZXJdLnZvdGVkLAogICAgICAgICAgICAiVGhlIHZvdGVyIGFscmVhZHkgdm90ZWQuIgogICAgICAgICk7CiAgICAgICAgcmVxdWlyZSh2b3RlcnNbdm90ZXJdLndlaWdodCA9PSAwKTsKICAgICAgICB2b3RlcnNbdm90ZXJdLndlaWdodCA9IDE7CiAgICB9CgogICAgLy8vIERlbGVnYXRlIHlvdXIgdm90ZSB0byB0aGUgdm90ZXIgYHRvYC4KICAgIGZ1bmN0aW9uIGRlbGVnYXRlKGFkZHJlc3MgdG8pIHB1YmxpYyB7CiAgICAgICAgLy8gYXNzaWducyByZWZlcmVuY2UKICAgICAgICBWb3RlciBzdG9yYWdlIHNlbmRlciA9IHZvdGVyc1ttc2cuc2VuZGVyXTsKICAgICAgICByZXF1aXJlKCFzZW5kZXIudm90ZWQsICJZb3UgYWxyZWFkeSB2b3RlZC4iKTsKCiAgICAgICAgcmVxdWlyZSh0byAhPSBtc2cuc2VuZGVyLCAiU2VsZi1kZWxlZ2F0aW9uIGlzIGRpc2FsbG93ZWQuIik7CgogICAgICAgIC8vIEZvcndhcmQgdGhlIGRlbGVnYXRpb24gYXMgbG9uZyBhcwogICAgICAgIC8vIGB0b2AgYWxzbyBkZWxlZ2F0ZWQuCiAgICAgICAgLy8gSW4gZ2VuZXJhbCwgc3VjaCBsb29wcyBhcmUgdmVyeSBkYW5nZXJvdXMsCiAgICAgICAgLy8gYmVjYXVzZSBpZiB0aGV5IHJ1biB0b28gbG9uZywgdGhleSBtaWdodAogICAgICAgIC8vIG5lZWQgbW9yZSBnYXMgdGhhbiBpcyBhdmFpbGFibGUgaW4gYSBibG9jay4KICAgICAgICAvLyBJbiB0aGlzIGNhc2UsIHRoZSBkZWxlZ2F0aW9uIHdpbGwgbm90IGJlIGV4ZWN1dGVkLAogICAgICAgIC8vIGJ1dCBpbiBvdGhlciBzaXR1YXRpb25zLCBzdWNoIGxvb3BzIG1pZ2h0CiAgICAgICAgLy8gY2F1c2UgYSBjb250cmFjdCB0byBnZXQgInN0dWNrIiBjb21wbGV0ZWx5LgogICAgICAgIHdoaWxlICh2b3RlcnNbdG9dLmRlbGVnYXRlICE9IGFkZHJlc3MoMCkpIHsKICAgICAgICAgICAgdG8gPSB2b3RlcnNbdG9dLmRlbGVnYXRlOwoKICAgICAgICAgICAgLy8gV2UgZm91bmQgYSBsb29wIGluIHRoZSBkZWxlZ2F0aW9uLCBub3QgYWxsb3dlZC4KICAgICAgICAgICAgcmVxdWlyZSh0byAhPSBtc2cuc2VuZGVyLCAiRm91bmQgbG9vcCBpbiBkZWxlZ2F0aW9uLiIpOwogICAgICAgIH0KCiAgICAgICAgLy8gU2luY2UgYHNlbmRlcmAgaXMgYSByZWZlcmVuY2UsIHRoaXMKICAgICAgICAvLyBtb2RpZmllcyBgdm90ZXJzW21zZy5zZW5kZXJdLnZvdGVkYAogICAgICAgIHNlbmRlci52b3RlZCA9IHRydWU7CiAgICAgICAgc2VuZGVyLmRlbGVnYXRlID0gdG87CiAgICAgICAgVm90ZXIgc3RvcmFnZSBkZWxlZ2F0ZV8gPSB2b3RlcnNbdG9dOwogICAgICAgIGlmIChkZWxlZ2F0ZV8udm90ZWQpIHsKICAgICAgICAgICAgLy8gSWYgdGhlIGRlbGVnYXRlIGFscmVhZHkgdm90ZWQsCiAgICAgICAgICAgIC8vIGRpcmVjdGx5IGFkZCB0byB0aGUgbnVtYmVyIG9mIHZvdGVzCiAgICAgICAgICAgIHByb3Bvc2Fsc1tkZWxlZ2F0ZV8udm90ZV0udm90ZUNvdW50ICs9IHNlbmRlci53ZWlnaHQ7CiAgICAgICAgfSBlbHNlIHsKICAgICAgICAgICAgLy8gSWYgdGhlIGRlbGVnYXRlIGRpZCBub3Qgdm90ZSB5ZXQsCiAgICAgICAgICAgIC8vIGFkZCB0byBoZXIgd2VpZ2h0LgogICAgICAgICAgICBkZWxlZ2F0ZV8ud2VpZ2h0ICs9IHNlbmRlci53ZWlnaHQ7CiAgICAgICAgfQogICAgfQoKICAgIC8vLyBHaXZlIHlvdXIgdm90ZSAoaW5jbHVkaW5nIHZvdGVzIGRlbGVnYXRlZCB0byB5b3UpCiAgICAvLy8gdG8gcHJvcG9zYWwgYHByb3Bvc2Fsc1twcm9wb3NhbF0ubmFtZWAuCiAgICBmdW5jdGlvbiB2b3RlKHVpbnQgcHJvcG9zYWwpIHB1YmxpYyB7CiAgICAgICAgVm90ZXIgc3RvcmFnZSBzZW5kZXIgPSB2b3RlcnNbbXNnLnNlbmRlcl07CiAgICAgICAgcmVxdWlyZShzZW5kZXIud2VpZ2h0ICE9IDAsICJIYXMgbm8gcmlnaHQgdG8gdm90ZSIpOwogICAgICAgIHJlcXVpcmUoIXNlbmRlci52b3RlZCwgIkFscmVhZHkgdm90ZWQuIik7CiAgICAgICAgc2VuZGVyLnZvdGVkID0gdHJ1ZTsKICAgICAgICBzZW5kZXIudm90ZSA9IHByb3Bvc2FsOwoKICAgICAgICAvLyBJZiBgcHJvcG9zYWxgIGlzIG91dCBvZiB0aGUgcmFuZ2Ugb2YgdGhlIGFycmF5LAogICAgICAgIC8vIHRoaXMgd2lsbCB0aHJvdyBhdXRvbWF0aWNhbGx5IGFuZCByZXZlcnQgYWxsCiAgICAgICAgLy8gY2hhbmdlcy4KICAgICAgICBwcm9wb3NhbHNbcHJvcG9zYWxdLnZvdGVDb3VudCArPSBzZW5kZXIud2VpZ2h0OwogICAgfQoKICAgIC8vLyBAZGV2IENvbXB1dGVzIHRoZSB3aW5uaW5nIHByb3Bvc2FsIHRha2luZyBhbGwKICAgIC8vLyBwcmV2aW91cyB2b3RlcyBpbnRvIGFjY291bnQuCiAgICBmdW5jdGlvbiB3aW5uaW5nUHJvcG9zYWwoKSBwdWJsaWMgdmlldwogICAgICAgICAgICByZXR1cm5zICh1aW50IHdpbm5pbmdQcm9wb3NhbF8pCiAgICB7CiAgICAgICAgdWludCB3aW5uaW5nVm90ZUNvdW50ID0gMDsKICAgICAgICBmb3IgKHVpbnQgcCA9IDA7IHAgPCBwcm9wb3NhbHMubGVuZ3RoOyBwKyspIHsKICAgICAgICAgICAgaWYgKHByb3Bvc2Fsc1twXS52b3RlQ291bnQgPiB3aW5uaW5nVm90ZUNvdW50KSB7CiAgICAgICAgICAgICAgICB3aW5uaW5nVm90ZUNvdW50ID0gcHJvcG9zYWxzW3BdLnZvdGVDb3VudDsKICAgICAgICAgICAgICAgIHdpbm5pbmdQcm9wb3NhbF8gPSBwOwogICAgICAgICAgICB9CiAgICAgICAgfQogICAgfQoKICAgIC8vIENhbGxzIHdpbm5pbmdQcm9wb3NhbCgpIGZ1bmN0aW9uIHRvIGdldCB0aGUgaW5kZXgKICAgIC8vIG9mIHRoZSB3aW5uZXIgY29udGFpbmVkIGluIHRoZSBwcm9wb3NhbHMgYXJyYXkgYW5kIHRoZW4KICAgIC8vIHJldHVybnMgdGhlIG5hbWUgb2YgdGhlIHdpbm5lcgogICAgZnVuY3Rpb24gd2lubmVyTmFtZSgpIHB1YmxpYyB2aWV3CiAgICAgICAgICAgIHJldHVybnMgKGJ5dGVzMzIgd2lubmVyTmFtZV8pCiAgICB7CiAgICAgICAgd2lubmVyTmFtZV8gPSBwcm9wb3NhbHNbd2lubmluZ1Byb3Bvc2FsKCldLm5hbWU7CiAgICB9Cn0=) works for me in Firefox.

So it will work at least for some people. The question would if it's enough of a problem to start looking for an alternative solution.

Reply:
Can you still add a warning here in case it is longer than 2000 characters?

Reply:
> Can do that easily from the user side.

Do you mean that just using `POST` from the client side should work? Does not seem to be the case. I checked now and looks like it's behind Cloudfront and `POST` requests do not even get through to it:

```bash
curl --request POST https://remix.ethereum.org/ --data "code=Y29udHJhY3QK&version=soljson-v0.8.4+commit.c7e474f2.js" --header "Host: remix.ethereum.org"
```
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<TITLE>ERROR: The request could not be satisfied</TITLE>
</HEAD><BODY>
<H1>403 ERROR</H1>
<H2>The request could not be satisfied.</H2>
<HR noshade size="1px">
This distribution is not configured to allow the HTTP request method that was used for this request. The distribution supports only cachable requests.
We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.
<BR clear="all">
If you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.
<BR clear="all">
<HR noshade size="1px">
<PRE>Generated by cloudfront (CloudFront)
Request ID: wNDoDUYB6ed6IWHQX9SVQJsn0hB715V4DKOvkdMTJIJ5pfeL2y2j7g==
</PRE>
<ADDRESS>
</ADDRESS>
</BODY></HTML>
```
Only `GET` requests work:
```
curl --request GET https://remix.ethereum.org --header "Host: remix.ethereum.org"
```

Reply:
> At the minimum you can output a warning on `make docs` if any snippet has an URL longer than 2000 bytes. We could then see if anything is affected.

ok, warning added. With the limit at 2000 bytes there are 32 snippets that do not pass it. With 4000 bytes it's 18 snippets. At 8000 bytes all the snippets pass.

Reply:
> > Can do that easily from the user side.
>
> Do you mean that just using POST from the client side should work? 

No, actually I meant to say `can't`, just made a typo.

Reply:
> Just ask remix to support lib compression and/or ascii85 (with a modified characters removing % and &) in the payload :)

Not sure if that's a serious comment but with all snippets being below 8k bytes maybe we could fit in 2k chars with good compression. Code is very squishy.

The snippet from `voting.rst` (one of the 3 longest) has 5415 bytes and just zipping it at default settings reduces it to 2055 bytes. The base64-encoded URL is a little bigger but not by all that much (7253 bytes for that snippet).

Reply:
> No, actually I meant to say can't, just made a typo.

Ah, ok :)

Anyway, given that this is a new feature and nobody is really relying on it yet - maybe it would be ok to go with a limit of 8k chars and call it a day? From the post you linked it looks like both Cloudfront and most of the popular browsers (except for old IE) should handle at least that much. We can always start looking for an alternative solution once we get many reports that these links do not work.